modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/Pipe.java

```diff
@@ -277,22 +288,33 @@ public ByteBuffer copyAndProduce(final ContentDecoder decoder) throws IOExceptio
         try {
             ByteBuffer duplicate = null;
             setInputMode(buffer);
+            int totalRead = 0;
             int bytesRead;
             try {
                 // clone original buffer
                 ByteBuffer originalBuffer = buffer.getByteBuffer();
-                bytesRead = decoder.read(originalBuffer);
+
+                // Drain the decoder until the end of stream is found or until the Pipe#buffer is full.
+                // bytesRead = -1 means reached out to the end of stream.
+                // bytesRead = 0 means Pipe's input buffer is full.
+                while ((bytesRead = decoder.read(buffer.getByteBuffer())) > 0) {
+                    totalRead += bytesRead;
+                }
                 duplicate = originalBuffer.duplicate();
                 // replicate positions of original buffer in duplicated buffer
                 int position = originalBuffer.position();
                 duplicate.limit(position);
-                if (bytesRead > 0) {
-                    duplicate.position(position - bytesRead);
+                if (totalRead > 0) {
+                    duplicate.position(position - totalRead);
                 }
             } catch (TruncatedChunkException ignore) {
```

Comment:
since here we are handling the TruncatedChunkException, let's rename the variable to ex (it's not ignored).

Reply:
Fixed in 1a225662c71267f47edbbee23cb3dcf7c2106fd3